### PR TITLE
feat(partition): flame and icicle performance and tweening

### PR DESCRIPTION
**Pull request by monfera**
## Summary

Helps the user orient where they are in data exploration by optionally adding brief animation. Increases render performance.

https://user-images.githubusercontent.com/1548516/109667342-5787c000-7b70-11eb-9dc6-a5a9354bd184.mp4

- increases performance on all partition charts that don't `showLegend` by not computing it
- adds optional (default: none) tween with configurable duration to the icicle/flame chart when using clipping [added](https://github.com/elastic/elastic-charts/pull/1033) by @dgieselaar 
- increases drilldown performance by deferring the reprojection (zoom/pan) and elimination of invisible boxes, using focus information in the renderer, a WebGL-compatible approach, therefore avoiding the need to recompute the viewmodel proper
- as the drilldown logic is different, the recently added `groupByRollup` and `tree.ts` additions for drilldown became unnecessary, they got removed
- optimizes the clipped text render path for icicle/flame for speed, by adding a specialized render function with reduced API and function call count, and by avoiding multirow text fill calculation in `viewmodel.ts`
- caches calls to time consuming `chroma` functions for a 2-3 OOM reduction of `chroma` calls
- moves commonly used functions eg. fill coloring out of `partition-chart`
- removes some `let` and conditionalization from the color logic

### Checklist

Delete any items that are not applicable to this PR.

- [X] This was checked for cross-browser compatibility
- [X] Proper documentation or storybook story was added for features that require explanation or tutorials
- [X] Unit tests were updated or added to match the most common scenarios

### CODE REVIEW

File: src/chart_types/partition_chart/layout/viewmodel/viewmodel.ts [hunk 1]
```diff
@@ -387,20 +411,27 @@ export function shapeViewModel(
     containerBackgroundColor,
   );
 
-  const pickQuads: PickFunction = (x, y) =>
-    quadViewModel.filter(
-      treemapLayout || icicleLayout || flameLayout
-        ? ({ x0, y0, x1, y1 }) => x0 <= x && x <= x1 && y0 <= y && y <= y1
-        : ({ x0, y0px, x1, y1px }) => {
+  const pickQuads: PickFunction = (x, y, { currentFocusX0, currentFocusX1 }) => {
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/viewmodel/viewmodel.ts:414]: I'll conditionalize it outside the loop, ie. `const pickQuads = sunburstLayout ? ... : ...`

**Review by markov00 — CHANGES_REQUESTED** [review]
I don't know exactly what happen but the drilldown no longer works in the flame chart and icicle chart examples in storybook. The drilldown is configured but didn't change the chart rendering:  ` drilldown: true` 
Seems that the svg highlighter get the drilldown, but not the canvas renderer

File: src/chart_types/partition_chart/renderer/canvas/canvas_linear_renderers.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ */
+
+import { clearCanvas } from '../../../../renderers/canvas';
+import { ShapeViewModel } from '../../layout/types/viewmodel_types';
+import { ContinuousDomainFocus } from './partition';
+
+const linear = (x: number) => x;
+const easeInOut = (alpha: number) => (x: number) => x ** alpha / (x ** alpha + (1 - x) ** alpha);
+
+let latestRaf: number;
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/renderer/canvas/canvas_linear_renderers.ts:27]: Thanks, will do!

File: src/chart_types/partition_chart/renderer/canvas/partition.tsx [hunk 3]
```diff
@@ -169,6 +186,7 @@ const mapDispatchToProps = (dispatch: Dispatch): ReactiveChartDispatchProps =>
 const DEFAULT_PROPS: ReactiveChartStateProps = {
   initialized: false,
   geometries: nullShapeViewModel(),
+  geometriesFoci: [],
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/renderer/canvas/partition.tsx:189]: just checked, `focuses` as plural is legit too, if you have some preference for the latter, glad to switch

**Comment by monfera** [comment]
Thanks for the catch @markov00 I added cc273f5d7caa2b5119764b3431db76715be9e842 to prevent drilldown on non-cropping versions

File: src/chart_types/partition_chart/renderer/canvas/canvas_linear_renderers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/renderer/canvas/canvas_linear_renderers.ts]: Done in ca4e1156f26d2ed03e0359cfb2edbe246c92e320

[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/partition_chart/renderer/canvas/canvas_linear_renderers.ts]: Thank you

**Review by markov00 — APPROVED** [review]
Tested locally, LGTM

**Pull request merged**